passing variable as "any" interface value causes error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: macos, macOS, 11.6, 20G165
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.0.0 (clang-1300.0.29.3)

getwd: /Users/nedpals/Documents/coding/vex
vmodules: /Users/nedpals/.vmodules
vroot: /Users/nedpals/Downloads/v
vexe: /Users/nedpals/Downloads/v/v
vexe mtime: 2022-03-21 08:28:02
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 d9cca53

Git version: git version 2.33.0
Git vroot status: weekly.2022.11-45-gd9cca53b
.git/config present: true
thirdparty/tcc status: thirdparty-unknown-unknown de82a130

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg bug.v`
```v
interface Any {}

fn return_any(val Any) ?Any {
    return val
}

fn main() {
    code := 200
    if an := return_any(code) {
        if an is int {
            println('an is an int $an')
        } else {
            println('an is not an int!')
        }
    }
}
```

**What did you expect to see?**

an is an int 200


**What did you see instead?**
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/bug.6991022573431803397.tmp.c:11351:42: error: implicit declaration of function 'I_int_literal_to_Interface_main__Any' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        if (_t1 = main__return_any(/*&main.Any*/I_int_literal_to_Interface_main__Any(&(*(code)))), _t1.state == 0) {
                                                ^
/tmp/v_501/bug.6991022573431803397.tmp.c:11351:42: note: did you mean 'I_int_to_Interface_main__Any'?
/tmp/v_501/bug.6991022573431803397.tmp.c:1917:25: note: 'I_int_to_Interface_main__Any' declared here
static inline main__Any I_int_to_Interface_main__Any(int* x) {
                        ^
/tmp/v_501/bug.6991022573431803397.tmp.c:11351:42: error: passing 'int' to parameter of incompatible type 'main__Any' (aka 'struct main__Any')
        if (_t1 = main__return_any(/*&main.Any*/I_int_literal_to_Interface_main__Any(&(*(code)))), _t1.state == 0) {
                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/bug.6991022573431803397.tmp.c:11342:61: note: passing argument to parameter 'val' here
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2045,7 +2045,7 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 		&& !expected_type.has_flag(.optional) {
 		if expr is ast.StructInit && !got_type.is_ptr() {
 			g.inside_cast_in_heap++
-			got_styp := g.cc_type(got_type.ref(), true)
+			got_styp := g.cc_type(got_type_raw.ref(), true)
 			// TODO: why does cc_type even add this in the first place?
 			exp_styp := exp_sym.cname
 			mut fname := 'I_${got_styp}_to_Interface_$exp_styp'
@@ -2056,12 +2056,7 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 				got_styp)
 			g.inside_cast_in_heap--
 		} else {
-			mut got_styp := g.cc_type(got_type, true)
-			got_styp = match got_styp {
-				'int' { 'int_literal' }
-				'f64' { 'float_literal' }
-				else { got_styp }
-			}
+			got_styp := g.cc_type(got_type_raw, true)
 			got_is_shared := got_type.has_flag(.shared_f)
 			exp_styp := if got_is_shared { '__shared__$exp_sym.cname' } else { exp_sym.cname }
 			// If it's shared, we need to use the other caster:

--- a/vlib/v/tests/cast_int_to_interface_test.v
+++ b/vlib/v/tests/cast_int_to_interface_test.v
@@ -1,0 +1,19 @@
+interface Any {}
+
+fn return_any(val Any) ?Any {
+	return val
+}
+
+fn test_cast_int_to_interface() {
+	code := 200
+	if an := return_any(code) {
+		if an is int {
+			println('an is an int!')
+		} else {
+			println('an is not an int!')
+		}
+		assert '$an' == 'Any(200)'
+	} else {
+		assert false
+	}
+}

